File: Java/Solution_6.java
public class Solution_6 {
    public String convert(String s, int numRows) {

        if (numRows == 1 || numRows >= s.length()) {
            return s;
        }

        StringBuilder[] rows = new StringBuilder[numRows];

        for (int i = 0; i < numRows; i++) {
            rows[i] = new StringBuilder();
        }

        int currentRow = 0;
        boolean goingDown = false;

        for (char c : s.toCharArray()) {
            rows[currentRow].append(c);

            if (currentRow == 0 || currentRow == numRows - 1) {
                goingDown = !goingDown;
            }

            currentRow += goingDown ? 1 : -1;
        }

        StringBuilder result = new StringBuilder();

        for (StringBuilder row : rows) {
            result.append(row);
        }

        return result.toString();
    }

    public static void main(String[] args) {
        Solution_6 solution = new Solution_6();

        String s1 = "PAYPALISHIRING";
        int numRows1 = 3;
        String s2 = "PAYPALISHIRING";
        int numRows2 = 4;
        String s3 = "A";
        int numRows3 = 1;

        System.out.println(solution.convert(s1, numRows1));  // Output: "PAHNAPLSIIGYIR"
        System.out.println(solution.convert(s2, numRows2));  // Output: "PINALSIGYAHRPI"
        System.out.println(solution.convert(s3, numRows3));  // Output: "A"
    }
}

File: Python/Solution_6.py
class Solution_6(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        # Edge case: if only one row or string length is less than the number of rows
        if numRows == 1 or numRows >= len(s):
            return s
        
        # Create an array of StringBuilder-like objects (lists) to hold the characters of each row
        rows = [''] * numRows
        
        currentRow = 0  # Start at the top row
        goingDown = False  # Variable to track direction
        
        # Traverse each character in the string
        for char in s:
            rows[currentRow] += char  # Append character to the current row
            
            # Change direction when hitting the top or bottom row
            if currentRow == 0 or currentRow == numRows - 1:
                goingDown = not goingDown
            
            # Update the row index (move up or down)
            currentRow += 1 if goingDown else -1
        
        # Combine all rows into one result string
        return ''.join(rows)
    
if __name__ == "__main__":
    solution = Solution_6()
    
    s = "PAYPALISHIRING"
    numRows = 3
    result = solution.convert(s, numRows)
    print("Input:", s)
    print("Number of Rows:", numRows)
    print("Converted Zigzag Output:", result)
    # Expected Output: "PAHNAPLSIIGYIR"